services/brig/test/integration/API/OAuth.hs

```diff
@@ -0,0 +1,865 @@
+      rt <- oatRefreshToken <$> createOAuthAccessToken brig accessTokenRequest
+      rid <- extractRefreshTokenId jwk rt
+      tokens <- C.runClient db (lookupOAuthRefreshTokens uid)
+      liftIO $ assertBool testMsg $ [rid3, rid] `hasSameElems` (oriId <$> tokens)
+  where
+    extractRefreshTokenId :: MonadIO m => JWK -> OAuthRefreshToken -> m OAuthRefreshTokenId
+    extractRefreshTokenId jwk rt = do
+      fromMaybe (error "invalid sub") . hcsSub <$> liftIO (verifyRefreshToken jwk (unOAuthToken rt))
+
+    hasSameElems :: (Eq a) => [a] -> [a] -> Bool
+    hasSameElems x y = null (x \\ y) && null (y \\ x)
+
+testRefreshTokenRetrieveAccessToken :: Brig -> Nginz -> Http ()
+testRefreshTokenRetrieveAccessToken brig nginz = do
+  uid <- userId <$> createUser "alice" brig
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [ReadSelf]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig uid scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  accessToken <- createOAuthAccessToken brig accessTokenRequest
+  get (nginz . paths ["self"] . authHeader (oatAccessToken accessToken)) !!! const 200 === statusCode
+  threadDelay $ 5 * 1000 * 1000 -- wait 5 seconds for access token to expire
+  get (nginz . paths ["self"] . authHeader (oatAccessToken accessToken)) !!! const 401 === statusCode
+  let refreshAccessTokenRequest = OAuthRefreshAccessTokenRequest OAuthGrantTypeRefreshToken cid secret (oatRefreshToken accessToken)
+  refreshedToken <- refreshOAuthAccessToken brig refreshAccessTokenRequest
+  get (nginz . paths ["self"] . authHeader (oatAccessToken refreshedToken)) !!! const 200 === statusCode
+
+testRefreshTokenWrongSignature :: Opts -> Brig -> Http ()
+testRefreshTokenWrongSignature opts brig = do
+  uid <- userId <$> createUser "alice" brig
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [ReadSelf]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig uid scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  accessToken <- createOAuthAccessToken brig accessTokenRequest
+  key <- liftIO $ readJwk (fromMaybe "path to jwk not set" (Opt.setOAuthJwkKeyPair $ Opt.optSettings opts)) <&> fromMaybe (error "invalid key")
+  badRefreshToken <- liftIO $ do
+    claims <- verifyRefreshToken key (unOAuthToken $ oatRefreshToken accessToken)
+    OAuthToken <$> signRefreshToken badKey claims
+  let refreshAccessTokenRequest = OAuthRefreshAccessTokenRequest OAuthGrantTypeRefreshToken cid secret badRefreshToken
+  refreshOAuthAccessToken' brig refreshAccessTokenRequest !!! do
+    const 403 === statusCode
+    const "Forbidden" === statusMessage
+
+testRefreshTokenNoTokenId :: Opts -> Brig -> Http ()
+testRefreshTokenNoTokenId opts brig = do
```

Comment:
```suggestion
testRefreshTokenNoTokens opts brig = do
```

(it's not about the token id, but about access and refresh token missing, no?)

Reply:
No, it's actually about a refresh token that does not contain an ID.